### DISCUSSION TITLE
group by 怎么动态分组？

### DISCUSSION THREAD

**Discussion by zhaoxingyu999** [dotnetcore/FreeSql#1038] [Q&A]
应用场景 ：
分组条件是a,b,c,d的任意组合，来自前端，前端选了 a就只分组a,选了 a,b就分组a,b
请问怎么用freesql写出来？
(select 部分也是来自前端，前端选了 a就只查a,选了 a,b就只查a,b
select  a,b,c,d,... from  xxx
group by a,b,c,d

**Comment by admintwgdh** [comment]
确实，我也遇到这个问题好多次了。

**Comment by admintwgdh** [comment]
要是作者可以完善一下就好了

**Comment by hd2y** [comment]
虽然不知道通过组装表达式树是否能实现，但是提供一个最简单最直接使用 `ToSql` 的方案，可以参考下：

```cs
string groupBy = "age,name";

var sql = freeSql.Select<Student>()
    .Where(t => t.Name.StartsWith("王"))
    .ToSql();

var list = freeSql.Select<object>()
    .WithSql($"select {groupBy} from ({sql}) tmp group by {groupBy}")
    .ToList<dynamic>("*");
```

对应会执行SQL：

```sql
SELECT *
FROM ( select age,name from (SELECT a."Id", a."Gender", a."Age", a."Name", a."Grade", a."Class"
    FROM "Student" a
    WHERE ((a."Name") LIKE '王%')) tmp group by age,name ) a
```

**Reply by zhaoxingyu999** [reply]
这种虽然能实现但是没有sum avg等聚合函数；
谢谢，我们已经找到另外一种方式，
就是通过添加判断方式 groupby null
```
        int c1 = 1;
            int c2 = 0;
            var tkdkdksql = g.sqlite.Select<TaskBuild>().From<Templates, Templates>((a, b, c) =>
             a.LeftJoin(aa => aa.TemplatesId == b.Id2 && b.Code == "xx")
             .LeftJoin(aa => aa.TemplatesId == c.Id2))
             .GroupBy((a, b, c) =>
                  new { K1 = c1 == 1 ? a.NamespaceName : null, K2 = c2 == 1 ? c.Code : null }
            ).ToSql((a) => new { K1 = a.Key.K1, K2 = a.Key.K2, Cnt = a.Count(), SumScore = a.Sum(a.Value.Item1.score ?? 0) });
```

是不成立就没有加入到groupby里面 显示的一个null列 不影响统计的
![image](https://user-images.githubusercontent.com/27735252/157799054-0c152c7b-91af-42f0-b0e7-a193ba256ea5.png)

**Comment by cqzxy** [comment]
.GroupBy(x => new
             {
                 O_ID = req.groupParams.djjg == 1 ? x.O_ID : null,
                 xzqh_id = req.groupParams.ywssqh == 1 ? x.xzqh_id : null,
                 qllx = req.groupParams.qllx == 1 ? x.qllx : null,
                 djlx = req.groupParams.djlx == 1 ? x.djlx : null
             })
             .ToList(x => new Dto
             {
                 djjg = req.groupParams.djjg == 1 ? x.Key.O_ID : null,
                 djlx = req.groupParams.djlx == 1 ? x.Key.djlx : null,
                 ywssqh =req.groupParams.ywssqh == 1 ? x.Key.xzqh_id : null,
                 qllx =req.groupParams.qllx == 1 ? x.Key.qllx : null,
                 bjsl = (int)x.Sum(x.Value.bjsl)
             });

**Reply by 2881099** [reply]
```c#
.GroupBy(x => new
{
    O_ID = req.groupParams.djjg == 1 ? x.O_ID : null,
    xzqh_id = req.groupParams.ywssqh == 1 ? x.xzqh_id : null,
    qllx = req.groupParams.qllx == 1 ? x.qllx : null,
    djlx = req.groupParams.djlx == 1 ? x.djlx : null
})
.ToList(x => new Dto
{
    djjg = x.Key.O_ID,
    djlx = x.Key.djlx,
    ywssqh = x.Key.xzqh_id,
    qllx = x.Key.qllx,
    bjsl = (int)x.Sum(x.Value.bjsl)
});
```

**Reply by leeolevis** [reply]
goooood

**Reply by admintwgdh** [reply]
重庆罩星宇？

**Reply by yinyihao1** [reply]
强烈建议写入文档 。 今天折腾好久

**Reply by 2881099** [reply]
> 强烈建议写入文档 。 今天折腾好久  

欢迎pr文档
